ui/src/org/pentaho/di/ui/cluster/dialog/SlaveServerDialog.java

```diff
@@ -525,35 +525,38 @@ public void ok() {
       if ( DialogUtils.objectWithTheSameNameExists( slaveServer, existingServers ) ) {
         String title = BaseMessages.getString( PKG, "SlaveServerDialog.SlaveServerNameExists.Title" );
         String message =
-            BaseMessages.getString( PKG, "SlaveServerDialog.SlaveServerNameExists", slaveServer.getName() );
+          BaseMessages.getString( PKG, "SlaveServerDialog.SlaveServerNameExists", slaveServer.getName() );
         String okButton = BaseMessages.getString( PKG, "System.Button.OK" );
         MessageDialog dialog =
-            new MessageDialog( shell, title, null, message, MessageDialog.ERROR, new String[] { okButton }, 0 );
+          new MessageDialog( shell, title, null, message, MessageDialog.ERROR, new String[] { okButton }, 0 );
 
         dialog.open();
         return;
       }
     }
 
-    originalServer.setName( slaveServer.getName() );
-    originalServer.setHostname( slaveServer.getHostname() );
-    originalServer.setPort( slaveServer.getPort() );
-    originalServer.setWebAppName( slaveServer.getWebAppName() );
-    originalServer.setUsername( slaveServer.getUsername() );
-    originalServer.setPassword( slaveServer.getPassword() );
+    synchronized ( originalServer ) {
```

Comment:
So what is being protected here by synchronizing at all ? This is code that only runs in a thick-client (Spoon) as it's the dialog. This code doesn't run on the server. So we're protecting someone that ostensibly starts a transformation or job using one or more SlaveServers, and then has edits the SlaveServer definition while the transformation/job is running. Seems to me that's irrational and they get what they get. That's a "don't do that".

With that behavior, I think the right answer is that the dialog should work solely on a clone of the SlaveServer. Only at the moment when the "Save" is performed should the underlying SlaveServer be mutated. But there's no knowing what will happen to the executing job/transformation.